firehose.go

```diff
@@ -269,11 +301,84 @@ func (fp *Firehose) RunLeader(ctx context.Context) {
 	}
 }
 
-// RunFirehose starts a Firehose running in leader of follower mode
-func (f *Firehose) RunFirehose() {
+func (fp *Firehose) IsLeader(ctx context.Context) (bool, error) {
+	nodeId := fp.StorageBackend.GetNodeId(ctx)
+	deploymentId := fp.StorageBackend.GetDeploymentId(ctx)
+	if nodeId == "" || deploymentId == "" {
+		panic("nodeId or deploymentId can not be determined")
+	}
+	leaderWriteErr := fp.StorageBackend.WriteLeaderFile(ctx, fp.processSettings.MetadataBucket, nodeId, deploymentId)
+
+	if leaderWriteErr == nil {
+		return true, nil
+	}
+	_, ok := leaderWriteErr.(LeaderFileExists)
+	if ok {
+		r, err := fp.StorageBackend.CreateReader(ctx, fp.processSettings.MetadataBucket, "leader.json")
+		if err != nil {
+			return false, err
+		}
+		data, err := ioutil.ReadAll(r)
+		if err != nil {
+			return false, err
+		}
+		var result LeaderFile
+		err = json.Unmarshal(data, &result)
+		if err != nil {
+			return false, err
+		}
+
+		if result.DeploymentId != deploymentId {
+			return false, fmt.Errorf("deploymentId does not match current leader")
```

Comment:
What happens during deploys? How does old leader file get deleted ?

Reply:
oh good point - on shutdown leader file has to be deleted right....  Although there is the situation of an ungraceful shutdown? Should we assume deploymentId is increasing and have it attempt to write a new leaderfile if it sees deploymentId in file is less than current?

Reply:
since we get to set the deploymentId right?

Reply:
We're using gae default for the deployment id but we can't assume it's order. In the design doc we said we should assume leader shuts down cleanly and in case it doesn't, you'd need to manually fix the dirty state. Let's go with that for now.

Reply:
ah okay will the context receive a signal on shutdown or I guess whats the cue for that?

Reply:
Yes can assume that context will get cancelled on shut down.